### DISCUSSION TITLE
GIF slows when rotate

### DISCUSSION THREAD

**Discussion by Flash-nDie** [libvips/libvips#3234] [Q&A]
Hi, I have a question in Vips compared to ImageMagick.
I use vips version 8.13.3

When I rotate GIF with magicksave, the speed is slower much than ImageMagick. 
But, when i resize GIF, the speed is similar to ImageMagick.
I used CLI:
Vips: vips rot "input.gif[n=-1]" output.gif d180 
ImageMagick7: convert -rotate 180 input.gif output.gif
The results with 460x250 GIF as belows

<img width="582" alt="image" src="https://user-images.githubusercontent.com/67739935/208568699-16890e59-e4a2-429e-820b-a426536162cd.png">
X-axis is frames and Y-axis is seconds.

When I tested the same image encoded into webp, The processing time, doing rotate or resize, looks half in vips compared to ImageMagick.(So fast)

I saw some Discussions, and the vips utilize ImageMagick in gif. But Why rotate is much slower?

**Comment by jcupitt** [comment]
Hi @Flash-nDie,

libvips always unpacks to RGBA on load, then reencodes to I (index) plus RGBA palette on save. Imagemagick always loads as I + RGBA and only unpacks to RGBA if it has to. 180 rotate can be done on just the index values so in this case IM can avoid regenerating the palette and redithering.

I see:

```
 /usr/bin/time -f %M:%e vips rot 3198.gif[n=-1] x.gif[reuse] d180
70328:1.09
$ /usr/bin/time -f %M:%e convert -rotate 180 3198.gif x.gif
96320:0.24
```

So IM is about 4x faster for this image (139 frames of 289x193 RGB), though it needs a bit more memory.

For resize, you do need to unpack to RGBA, so they are doing the same work. I see:

```
$ /usr/bin/time -f %M:%e convert huge2.gif -thumbnail 128x128 x.gif
728452:2.60
$ /usr/bin/time -f %M:%e vipsthumbnail huge2.gif[n=-1] -o x.gif
238380:1.55
```

Now libvips is maybe 40% faster, and needs 1/3 or the memory.

**Comment by jcupitt** [comment]
> I saw some Discussions, and the vips utilize ImageMagick in gif. 

libvips has had its own GIF write for a while and no longer needs imagemagick. You'll see the best performance with libvips configured to use cgif for GIF save.

**Reply by Flash-nDie** [reply]
@jcupitt 
Thank you for answering. 
As you pointed out, I used cgif also, but the trend was similar when using cgifs.
However, from what you explained, I found out why vips took much.
One thing, is there any material or content that I can refer to what the index is about? I know RGBA, but I don't know what index means. Is it only necessary to modify this index in Rotate?
I understood that unpack is transforming image binary(RGBA + I ) to pixel array(cache in vips or IM), and IM doesn't transform the binary to pixel. is that right?

**Reply by jcupitt** [reply]
It's indexed colour, eg.: https://en.wikipedia.org/wiki/Indexed_color
